modelzoo/admin.py

```diff
@@ -0,0 +1,475 @@
+        else:
+            self.token = token
+
+        self.adminF = admin
+        self.admin_token = None
+        if admin:
+            self.admin_token = conn._authenticate(self.username, self.password)
+        self.parent = None
+
+    def make(
+        self,
+        conn: ModelZooConnection,
+        username: typing.Optional[str] = "",
+        password: typing.Optional[str] = "",
+        token: typing.Optional[str] = "",
+    ) -> UserAuth:
+        if self.username == username and username != "":
+            if self.password == password and password != "":
+                if self.token == token and token != "":
+                    return self
+                else:
+                    u = UserAuth(
+                        conn,
+                        username=username,
+                        password=password,
+                        token=token,
+                        admin=False,
+                    )
+                    u.parent = self
+                    return u
+        u = UserAuth(conn, username=username, password=password, token=token)
+        return u
+
+    @property
+    def admin(self):
+        if self.adminF:
+            return self.admin_token
+        if self.parent is None:
+            return self.token
+        return self.parent.admin
+
+
+class ModelZooConnection(object):
+    def __init__(
+        self,
+        address: typing.Optional[str] = "grpc.modelzoo.live",
+        token: typing.Optional[str] = "",
+        username: typing.Optional[str] = "",
+        password: typing.Optional[str] = "",
+    ):
+        self.conn = None
+        self.address = address
+        self.conn_error = ModelZooConnectionException(
+            "Must be connected to ModelZoo instance before issuing requests. Please call connect()."
+        )
+        try:
+            self.auth = UserAuth(
+                self, username=username, password=password, token=token
+            )
+        except AuthenticationException as e:
+            self.auth = None
+
+    def connect(self, address: typing.Optional[str] = "grpc.modelzoo.live"):
+        self.address = self.address if address is None else address
+        channel = grpc.insecure_channel(self.address)
+        self.conn = ModelStub(channel)
+
+    def authenticate(
+        self,
+        username: typing.Optional[str] = "",
+        password: typing.Optional[str] = "",
+        token: typing.Optional[str] = "",
+    ) -> None:
+        try:
+            self.auth = UserAuth(
+                self, username=username, password=password, token=token
+            )
+        except InsufficientCredentialsException:
+            raise ModelZooConnectionException(
+                "At least one of [(Username, Password), Token] are required."
+            )
+        except InvalidTokenException:
+            raise ModelZooConnectionException("The token provided is not valid.")
+
+    def _authenticate(self, username: str, password: str) -> str:
```

Comment:
I see. This `_authenticate` is actually calling the authenticate endpoint. can we put a more descriptive name in here?